### PR TITLE
Add galaxy tool for cufflinks gffread command.

### PR DESCRIPTION
**Pull request by jj-umn**
The tool parameters are organized so that only valid combinations of parameters should be possible.

### CODE REVIEW

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+    #end if
+    #if $maxintron and $maxintron > 0:
+        -i $maxintron
+    #end if
+    #if $region.region_filter == 'filter':
+        -r $region.range $region.discard_partial
+    #end if
+    #if $merging.merge_sel != 'none':
+        $merging.merge_cmd
+        #echo ' '.join(str($merging.merge_options).split(','))
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: I hope to make such statements a little bit simpler with: https://bitbucket.org/galaxy/galaxy-central/pull-request/624/add-__getitem__-to-select-parameters-to

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+            #end if
+            $gffs.output_cmd
+        #elif $gffs.gff_fmt == 'gtf':
+            $gffs.output_cmd
+        #end if
+    #end if
+]]>
+    </command>
+    <inputs>
+        <param name="input" type="data" format="gff3,gtf" label="Input GFF3 or GTF feature file"/>
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: How good is our gff3 sniffer? Maybe we need to add "gff" as well?

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+            <when value="filter">
+                <param name="range" type="text" value="" size="60" label="Only show transcripts overlapping coordinate range"
+                    help="-r [['strand']'chr':]'start'..'end' &lt;br&gt; examples: &lt;br&gt; 1000..500000 &lt;br&gt; chr1:1000..500000 &lt;br&gt; +chr1:1000..500000 &lt;br&gt; -chr1:1000..500000" >
+                    <validator type="regex">(([+-])?(\w+:))?\d+\.\.\d+</validator>
+                </param>
+                <param name="discard_partial" type="boolean" truevalue="-R" falsevalue="" check="false" 
+                       label="(-R) and discard all transcripts that are not fully contained within the given range"/>
+            </when>
+        </conditional>
+        <param name="maxintron" type="integer" value="" optional="true" min="0" label="(-i) max_intron - Filter out transcipts with large introns"
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: @jj-umn can we add the (-i) to the end of help?

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+                    help="-r [['strand']'chr':]'start'..'end' &lt;br&gt; examples: &lt;br&gt; 1000..500000 &lt;br&gt; chr1:1000..500000 &lt;br&gt; +chr1:1000..500000 &lt;br&gt; -chr1:1000..500000" >
+                    <validator type="regex">(([+-])?(\w+:))?\d+\.\.\d+</validator>
+                </param>
+                <param name="discard_partial" type="boolean" truevalue="-R" falsevalue="" check="false" 
+                       label="(-R) and discard all transcripts that are not fully contained within the given range"/>
+            </when>
+        </conditional>
+        <param name="maxintron" type="integer" value="" optional="true" min="0" label="(-i) max_intron - Filter out transcipts with large introns" 
+               help="If set, discard transcripts having an intron larger"/>
+        <param name="chr_replace" type="data" format="tabular" optional="true" label="Replace reference sequence names (e.g. chr1 with 1)" >
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: (e.g. ...) is also help.

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+            <help>(-m chr_replace) &lt;br&gt; 
+                  chr_replace is input file is a 2 column tab-delimited file containing a reference (genomic) sequence replacement table with this format: &lt;br&gt; 
+                  "original_ref_ID"  "new_ref_ID" &lt;br&gt; 
+                  GFF records on reference sequences that are not found among the "original_ref_ID" entries in this file will be filtered out
+            </help>
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: Ha, this is a nice idea for large help texts. We can also use the CDATA tag here as well.

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+                   (-s seq_info.fsize -A)  useful with mRNA/EST/protein mappings &lt;br&gt;
+                  seq_info input file is a 3 column tab-delimited file providing this info for each of the mapped sequences: &lt;br&gt;
+                  "seq-name" "seq-length" "seq-description" &lt;br&gt;
+            </help>
+        </param> 
+        -->
+
+        <!-- merging -->
+        <conditional name="merging">
+            <param name="merge_sel" type="select" label="(-M) Transcript merging">
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: (-M) should go to the help section.

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+            </when>
+            <when value="cluster">
+                <param name="merge_cmd" type="hidden" value="--cluster-only"/>
+                <expand macro="cluster_opt_sel" />
+            </when>
+        </conditional>
+        <!-- reference sequence file -->
+        <!-- Error: -g option is required for options -w, -x, -y, -V, -N, -M -->
+        <conditional name="reference_genome">
+            <param name="source" type="select" label="(-g) Reference Genome (Required for fasta outputs)">
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: (-g) to help

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+            <when value="history">
+                <param name="genome_fasta" type="data" format="fasta" label="Genome Reference Fasta"/> 
+                <expand macro="ref_filtering_select" />
+                <expand macro="fasta_output_select" />
+            </when>
+        </conditional>
+
+        <!-- outputs -->
+        <conditional name="gffs">
+            <param name="gff_fmt" type="select" optional="true" label="(-o) Feature File Output">
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: (-o)

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+                <option value="none">none</option>
+                <option value="gff">GFF</option>
+                <option value="gtf">GTF</option>
+            </param>
+            <when value="none">
+            </when>
+            <when value="gff">
+                <param name="output_cmd" type="hidden" value="-o output.gff3"/>
+                <param name="ensembl" type="boolean" truevalue="-F" falsevalue="" check="false" label="(-L) Ensembl GTF to GFF3 conversion"/>
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: (-L)

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+            <param name="ref_filtering" type="select" display="checkboxes" multiple="true" label="reference based filters">
+                <option value="-N">(-N) discard multi-exon mRNAs that have any intron with a non-canonical splice site consensus (i.e. not GT-AG, GC-AG or AT-AC)</option>
+                <option value="-J">(-J) discard any mRNAs that either lack initial START codon or the terminal STOP codon, or have an in-frame stop codon (only print mRNAs with a fulll, valid CDS)</option>
+                <option value="-V">(-V) discard any mRNAs with CDS having in-frame stop codons</option>
+                <option value="-H">(-H with -V) check and adjust the starting CDS phase if the original phase leads to a translation with an in-frame stop codon</option>
+                <option value="-B">(-B with -V) single-exon transcripts are also checked on the opposite strand</option>
+            </param>
+        </xml>
+        <xml name="trackname">
+            <param name="tname" type="text" value="" size="30" optional="true" label="(-t) Trackname to use in the second column of each GFF output line">
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: (-t) should be moved to help

File: tool_collections/cufflinks/gffread/gffread.xml [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+      (use -o- for printing to stdout)
+  -t  use <trackname> in the second column of each GFF output line
+  -T  -o option will output GTF format instead of GFF3
+
+
+
+
+
+]]>
+    </help>
```
Comment (bgruening) [inline tool_collections/cufflinks/gffread/gffread.xml]: Citation macro is missing here.

**Comment by bgruening** [comment]
Very nice wrapper! I love the tests!

**Comment by jj-umn** [comment]
Thanks for the careful review.   

So in general:  any reference to an option name in a param should go in the help tag.  

With options within a select param, should the option name go at the start or end of the description:
<param name="ref_filtering" type="select" display="checkboxes" multiple="true" label="reference based filters">
          <option value="-N">(-N) discard multi-exon mRNAs that have any intron with a non-canonical splice site consensus (i.e. not GT-AG, GC-AG or AT-AC)</option>
OR?
          <option value="-N">(discard multi-exon mRNAs that have any intron with a non-canonical splice site consensus (i.e. not GT-AG, GC-AG or AT-AC) (-N) </option>

**Comment by bgruening** [comment]
I usually put them at the end of a select-text. So second option :)

**Comment by nsoranzo** [comment]
I also put the option name at the end of the option content.

**Comment by jj-umn** [comment]
Commited new gffread.xml with suggested changes.

**Comment by bgruening** [comment]
Thanks @jj-umn. Sorry that it took so long.

**Pull request merged**